### PR TITLE
Functions: S4 in-memory storage

### PR DESCRIPTION
**Pull request by pinebit**

### CODE REVIEW

**Review by bolekk — CHANGES_REQUESTED** [review]

File: core/services/s4/envelope.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"crypto/ecdsa"
+	"encoding/json"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// Envelope represents a JSON object that is signed for address verification.
+// All []byte values are encoded as base64 (default JSON behavior).
+// Hex is not used to avoid confusion due to case-sensivity and 0x prefix.
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/envelope.go:13]: That json is not meant to be read. It is temporary constructed in runtime for both client and server and should not even appear in logs.

File: core/services/s4/storage.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	Payload []byte
+	// Version attribute assigned by user
+	Version int64
+	// Expiration timestamp assigned by user (milliseconds)
+	Expiration int64
+}
+
+// Metadata is the internal S4 data associated with a Record
+type Metadata struct {
+	Confirmed         bool
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/storage.go:43]: Initially I did that. But then I realized that "EXPIRED" can easily be deducted from `Expiration < now()`. Also, changing records state to EXPIRED would require additional write operations on DB, whereas a simply query `where expiration < now()` would do the trick.
Maybe I will revisit this when working on the plugin. But for now, the simpler the better.

File: core/services/s4/storage.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	"errors"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type RecordState int
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/storage.go:14]: A leftover, removed.

File: core/services/s4/envelope.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		SlotID:     slotId,
+		Payload:    record.Payload,
+		Version:    record.Version,
+		Expiration: record.Expiration,
+	}
+}
+
+// Sign calculates signature for the serialized envelope data.
+func (e Envelope) Sign(privateKey *ecdsa.PrivateKey) (signature []byte, err error) {
+	js, err := json.Marshal(e)
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/envelope.go:34]: The current implementation (standard) of `JSON.stringify` and `json.Marshal` respect declaration order and do not insert any whitespace unless requested by the user. If there is a system that has a different behavior, then it needs to implement custom json serialization. I

File: core/services/s4/storage.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil && !errors.Is(err, ErrEntryNotFound) {
+		return err
+	}
+
+	highestExpiration := record.Expiration
+	if entry != nil {
+		highestExpiration = entry.HighestExpiration
+		if highestExpiration < record.Expiration {
+			highestExpiration = record.Expiration
+		}
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/storage.go:133]: The new record expiration can be larger than the previously recorded highest expiration, right?
I feel we have misunderstanding on the purpose of highest expiration term.
I think this "highest expiration" complicates the things, I will get back to the CLIP to find a better idea.

File: core/services/s4/orm.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+)
+
+var (
+	ErrEntryNotFound = errors.New("entry not found")
+)
+
+// Entry represents a data row persisted by ORM.
+type Entry struct {
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/orm.go]: Fixed

File: core/services/s4/envelope.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		Address:    address.Bytes(),
+		SlotID:     slotId,
+		Payload:    record.Payload,
+		Version:    record.Version,
+		Expiration: record.Expiration,
+	}
+}
+
+// Sign calculates signature for the serialized envelope data.
+func (e Envelope) Sign(privateKey *ecdsa.PrivateKey) (signature []byte, err error) {
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/envelope.go:35]: See my PM. I think we need to get rid of HighestExpiration at all and make GC mandatory. NOPs shall not take risks while persisting any user records for a long time. They only need to guarantee persistence until Expiration deadline.

[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/envelope.go:37]: UPD: removed HighestExpiration. The user queries data by address/slot/version.

File: core/services/s4/envelope.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+func (e Envelope) ToJson() ([]byte, error) {
+	address, err := json.Marshal(e.Address)
+	if err != nil {
+		return nil, err
+	}
+	payload, err := json.Marshal(e.Payload)
+	if err != nil {
+		return nil, err
+	}
+	js := fmt.Sprintf(`{"address":%s,"slotid":%d,"payload":%s,"version":%d,"expiration":%d}`, address, e.SlotID, payload, e.Version, e.Expiration)
```
[the remark this answers is not in the document]
Comment (bolekk) [reply core/services/s4/envelope.go:69]: Hopefully we can silence those by some clever string concatenation...?

**Review by bolekk — APPROVED** [review]
LGTM

File: core/services/s4/storage.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+// Record represents a user record persisted by S4.
+type Record struct {
+	// Arbitrary user data
+	Payload []byte
+	// Expiration timestamp assigned by user (unix time in milliseconds)
+	Expiration int64
+}
+
+// Metadata is the internal S4 data associated with a Record
+type Metadata struct {
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/storage.go:39]: `Storage` interface is not meant to be used by OCR plugin, it is a public "user" interface. OCR plugin will use ORM.

File: core/services/s4/storage.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil || signer != key.Address {
+		return ErrWrongSignature
+	}
+
+	row, err := s.orm.Get(key.Address, key.SlotId, pg.WithParentCtx(ctx))
+	if err != nil && !errors.Is(err, ErrNotFound) {
+		return err
+	}
+
+	if row != nil && key.Version <= row.Version {
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/storage.go:135]: `Storage` interface is not meant to be used by OCR plugin, it is a public "user" interface. OCR plugin will use ORM.

File: core/services/s4/orm.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	// If a row is missing, ErrNotFound is returned.
+	// Returned row is a clone, safe to modify.
+	Get(address common.Address, slotId uint, qopts ...pg.QOpt) (*Row, error)
+
+	// Put inserts (or updates) a row identified by the specified address and slotId.
+	// No validation is applied for signature, version, etc.
+	// Implementation clones the given Row when persisting.
+	Upsert(address common.Address, slotId uint, row *Row, qopts ...pg.QOpt) error
+
+	// DeleteExpired deletes any entries having HighestExpiration < now().
```
[the remark this answers is not in the document]
Comment (pinebit) [reply core/services/s4/orm.go:33]: Good catch, will fix it in the next PR

**Comment by pinebit** [comment]
Merging to proceed with OCR plugin, I will address the remaining minor comments in the next PR.

**Pull request merged**